KEP: hardened exec requests (#1899)
kubernetes/enhancements: keps/sig-node/1898-hardened-exec/README.md

#### 2. Require a `POST` request to the streaming endpoints

Exec, attach, and port-forward (and run) all respond to either GET or POST requests. Currently the
kube-apiserver uses the same request method when calling the Kubelet as the incoming client request
(e.g. a GET exec request to the apiserver results in a GET exec request to the kubelet). The API
server will need to be updated to always POST to these endpoints, regardless of the incoming client
request.

Since we do not support Kubelets newer than the kube-apiserver, we can safely remove support for GET
requests to these endpoints in the same release that the apiserver moves to unconditional POST.

Although attach & port-forward do not have the same risks of arbitrary code execution as exec, they
share a lot of the same code and should match the exec logic to reduce complexity.

#### 3. Remove the `UID` versions of the endpoints

The exec, attach, and port-forward handlers all support 2 URL path formats:

1. `/exec/{podNamespace}/{podID}/{containerName}`
2. `/exec/{podNamespace}/{podID}/{uid}/{containerName}`

The second path format includes the pod's UID, and is never called by the kube-apiserver. We should

> remove these endpoints as they're unused, to reduce code complexity.

Reviewer: Aside: if you wanted to preserve the UID checking behavior, it might be feasible to make up a conditional header `If-UID:` and fail requests where that's set and not the same as the running Pod's UID.

Author: I don't think it's used anywhere. It would be easy enough to add this approach later if a use case arose.

- If only body options are provided, use those.
- If only query options are provided, use those.
- If both body and query options are provided, require that they be identical.

Eventually, body options can be required. The requirement will be controlled by the
`HardenedExecRequests` feature gate. The feature will remain in alpha for TBD releases after
supported clients have been updated with the new request format.

<<[UNRESOLVED]>>

OPEN QUESTION: How many releases should we wait before requiring request body parameters?

<<[/UNRESOLVED]>>

To minimize risk of breakage, attach and port-forward requests should be left unchanged, as they do
not share the same risks of arbitrary code execution as exec.

#### 2. Match new Kubelet request requirements.

The API server must be updated to unconditionally send POST requests to the Kubelet (for
exec/attach/port-forward), and provide the request options in the body. See [2. Require a
<code>POST</code> request to the streaming
endpoints](#2-require-a--request-to-the-streaming-endpoints) and [4. Require the request options to
be provided in the request body](#4-require-the-request-options-to-be-provided-in-the-request-body).

> #### 3. Require the websocket protocol for GET requests.

Reviewer: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#websockets-and-spdy says:
> Clients should use the SPDY protocols if their clients have native support, or WebSockets as a fallback.
> …
> In the future, an HTTP/2 implementation will be exposed that deprecates SPDY.

Is that still accurate?

Reviewer: I'm not clear on why some clients (browsers?) aren't able to use POST + WebSocket - it'd be nice to have that clarified.

Reviewer 2: correct. browsers don't allow controlling the http method when opening a websocket

Reviewer: I think I get it. If I'm right, the challenge here is that the client sends a single request to the API server that is both an upgrade request _and_ request for a command execution. But I'm not sure I've understood, even after re-reading several times.

I haven't looked at the API here because I don't think there's documentation on how this works. The upgrade to WebSocket includes the exec request in its query string. So the options could include:
1. Retain current behavior because it's not vulnerable to this attack
2. POST requesting an exec, API server returns a WebSocket URI, client send a (GET) `Upgrade:` to open that WebSocket, continue conversation over WebSocket
3. Client requests a WebSocket, upgrade happens, and then within the WebSocket stream request an exec

I couldn't be sure I've understood what this section is calling for. With that in mind, would you be willing @tallclair to reword this to clarify?

Author: 1. It is vulnerable to attack, but by locking it down to only websocket requests, it makes it harder to exploit (especially in environments where websockets aren't used).
2. This is similar to the approach we previously took for the apiserver <-> CRI communications. I like the idea, but it's also a breaking change. We could skip the caching layer that we used in the CRI implementation by just signing the request parameters (i.e. POST request to get the signed query params, follow up with the GET including the signature)
3. I'm not familiar enough with the websocket protocol to know how complicated this would be. It's certainly feasible, but as it's also a breaking change I'm not sure how it compares with #2

I'll update the KEP with these alternatives.

Author: Updated with these options.

Reviewer 2: What does a browser do if it is redirected from an https://... endpoint to a wss://... endpoint? Does it open a websocket request?

Reviewer: `GET` (and other HTTP verbs) don't have any meaning for WebSockets, so I think the browser will error out.
(untested)

Author: I haven't tested it, but the redirected request would be missing the required headers for the opening handshake, so I assume it wouldn't work. https://tools.ietf.org/html/rfc6455#section-1.3

Author: Thinking about this more, I have a strong preference for "alternative 1" (hash of request as a subprotocol). Alternative 2 (exec token request) would require managing a signing key and option 3 (make request over the websocket connection) would be a much larger change. Since headers can't be modified on a redirect, I think the hash approach provides sufficient protection with minimal changes on the client & server side. We just need a well-defined serialized input for the hash function. Then the question is whether we want to tie this to the HardenedExecRequests feature gate.

to be referenced in the body mitigates vulnerabilities that reuse the request body (e.g. a 307
redirect). For backwards compatibility, during the transition period the kube-apiserver will need to
provide the options in both the body and the URL. Newer Kubelets should verify the options match,
and older Kubelets will just ignore the body.

### API Server Changes

The changes in the apiserver are:

1. Require options to be included in the request body for POST exec requests.
2. Match new kubelet request requirements.
3. Require the websocket protocol for GET requests.

#### 1. Require options to be included in the POST request body for exec requests.

The kube-apiserver should read options from the POST request body for exec requests. The pod
reference should be added for the same reasons as described in [4. Require the request options to be
provided in the request body](#4-require-the-request-options-to-be-provided-in-the-request-body).

Initially, this should be optional for backwards compatibility:

- If only body options are provided, use those.
- If only query options are provided, use those.

> - If both body and query options are provided, require that they be identical.

Author: The current behavior completely ignores the request body. Only the query parameters are used.

<<[/UNRESOLVED]>>

<<[UNRESOLVED]>>

Alternative options for protecting websockets at the expense of introducing a breaking change:

1. As extra protection against redirect attacks, we could use a similar trick as is used for
providing the bearer token as an extra sub-protocol:
https://github.com/kubernetes/kubernetes/blob/21953d15ea48972f20a8de29d58bd5ce6d913914/staging/src/k8s.io/apiserver/pkg/authentication/request/websocket/protocol.go#L37-L38
In this case, we could take a hash of the request options and provide the hash as an additional
sub-protocol. Then, the exec handler would need to validate that the request options matched the
hash value in the sub-protocol.

2. Approve the exec params with a separate POST request. For example, the client issues a POST to
   `.../exec/token` with the exec params included in the request body. The response includes a token
   that must be included in the GET exec params. Implementation options include caching a single-use
   token (similar to how CRI streaming requests work), or signing the request params.

3. The initial websocket request only opens the websocket protocol stream, and a subsequent request

>    must be sent over the websocket to initiate the actual GET request.

Reviewer: ```suggestion
    must be sent over the websocket to initiate the actual exec or other action.
```
(because WebSocket, as a byte stream, doesn't have HTTP verbs)

## Proposal

### Kubelet API

The changes to the Kubelet API are:

1. Remove the `/run` endpoint, which is unused in core Kubernetes.
2. Require a `POST` request for the `/exec`, `/attach` and `/portForward` endpoints.
3. Remove the `/{UID}/` versions of the endpoints.
4. Require the exec/attach/port-forward options and pod reference to be provided in the request body.

**Note on backwards compatibility:** This proposal assumes that the only client that should be
talking to these endpoints on the Kubelet is the kube-apiserver. Therefore, while the changes must
support API server version skew, endpoints and request types that are unused by the kube-apiserver
can be removed immediately. The [version skew
policy](https://kubernetes.io/docs/setup/release/version-skew-policy/#kubelet) requires that the
Kubelet not be newer than the kube-apiserver, so backwards compatibility with older apiservers is
not required.

#### 1. Remove the `/run` endpoint

The `run` endpoint provides the option to run a command in a container with a synchronous
request. This endpoint is not used by any Kubernetes components, and should be removed to reduce the
attack surface.

> **Risk:** Although this endpoint is not exposed in the Kubernetes API, it is still reachable via a

Reviewer 3: I am inclined to just remove it.  We gave no version guarantee on the endpoint.
